datafusion/optimizer/src/replace_distinct_aggregate.rs

```diff
@@ -32,6 +36,22 @@ use datafusion_expr::{Aggregate, Distinct, LogicalPlan};
 /// ```text
 /// SELECT a, b FROM tab GROUP BY a, b
 /// ```
+///
+/// On the other hand, for a `DISTINCT ON` query the replacement is
+/// a bit more involved and effectively converts
+/// ```text
+/// SELECT DISTINCT ON (a) b FROM tab ORDER BY a DESC, c
+/// ```
+///
+/// into
+/// ```text
+/// SELECT b FROM (
+///     SELECT a, FIRST_VALUE(b ORDER BY a DESC, c) AS b
```

Comment:
```suggestion
///     SELECT a, FIRST_VALUE(PARTITION BY b ORDER BY a DESC, c) AS b
```

Reply:
Actually there is no partitioning involved in the transformed plans. In fact the query from docs returns the same results as the `DISTINCT ON` query:
```sql
❯ select c2 from (
        select c1, first_value(c2 order by c1, c3) as c2 
        from test group by c1
  ) 
  order by c1;
+----+
| c2 |
+----+
| 5  |
| 4  |
| 2  |
| 1  |
| 3  |
+----+
5 rows in set. Query took 0.035 seconds.

❯ select distinct on (c1) c2 from test order by c1, c3;
+---------+
| test.c2 |
+---------+
| 5       |
| 4       |
| 2       |
| 1       |
| 3       |
+---------+
5 rows in set. Query took 0.034 seconds
```

Reply:
Sorry -- thank you for the clarification